keystore/keystore.go

```diff
@@ -0,0 +1,222 @@
+	switch keyType {
+	case Ed25519:
+		return ed25519.PublicKey(internal.Bytes(privateKeyBytes)), nil
+	case EcdsaSecp256k1:
+		// Here we use SEC1 (uncompressed) format for ECDSA public keys.
+		// Its commonly used and EVM addresses are derived from this format.
+		// We use the geth crypto library for secp256k1 support
+		// because stdlib doesn't support it.
+		privateKey, err := gethcrypto.ToECDSA(internal.Bytes(privateKeyBytes))
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert private key to ECDSA private key: %w", err)
+		}
+		pubKey := gethcrypto.FromECDSAPub(&privateKey.PublicKey)
+		return pubKey, nil
+	case X25519:
+		rv, err := curve25519.X25519(internal.Bytes(privateKeyBytes)[:], curve25519.Basepoint)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive shared secret: %w", err)
+		}
+		return rv, nil
+	default:
+		// Some types may not have a public key.
+		return []byte{}, nil
+	}
+}
+
+type keystore struct {
+	mu       sync.RWMutex
+	keystore map[string]key
+	storage  storage.Storage
+	enc      EncryptionParams
+}
+
+func NewKeystore(ctx context.Context, storage storage.Storage, enc EncryptionParams) (Keystore, error) {
+	ks, err := load(ctx, storage, enc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load keystore: %w", err)
+	}
+	return &keystore{
+		mu:       sync.RWMutex{},
+		keystore: ks,
+		storage:  storage,
+		enc:      enc,
+	}, nil
+}
+
+func load(ctx context.Context, storage storage.Storage, enc EncryptionParams) (map[string]key, error) {
+	encryptedKeystore, err := storage.GetEncryptedKeystore(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get encrypted keystore: %w", err)
+	}
+
+	// If no data exists, return empty keystore
+	if encryptedKeystore == nil || len(encryptedKeystore) == 0 {
+		return make(map[string]key), nil
+	}
+
+	encryptedSecrets := gethkeystore.CryptoJSON{}
+	err = json.Unmarshal(encryptedKeystore, &encryptedSecrets)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal encrypted keystore: %w", err)
+	}
+	decryptedKeystore, err := gethkeystore.DecryptDataV3(encryptedSecrets, enc.Password)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt keystore: %w", err)
+	}
+	keystorepb := &serialization.Keystore{}
+	err = proto.Unmarshal(decryptedKeystore, keystorepb)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal keystore: %w", err)
+	}
+	keystore := make(map[string]key)
+	for _, k := range keystorepb.Keys {
+		pkRaw := internal.NewRaw(k.PrivateKey)
+		keyType := KeyType(k.KeyType)
+		publicKey, err := publicKeyFromPrivateKey(pkRaw, keyType)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get public key from private key: %w", err)
+		}
+		keystore[k.Name] = newKey(keyType, pkRaw, publicKey, time.Unix(k.CreatedAt, 0), k.Metadata)
+	}
+	return keystore, nil
+}
+
+func save(ctx context.Context, storage storage.Storage, enc EncryptionParams, keystore map[string]key) error {
```

Comment:
Did you consider making load/save methods on `keystore`? Or does the map copying make it odd?

Reply:
wasn't really map copying but I think I was thinking this would be a bit more readable (esp in tests). But on second thought I think its fine I can just do that